module/block_iterator.go

```diff
@@ -0,0 +1,114 @@
+package module
+
+import (
+	"fmt"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// IterateJob defines the range of blocks to iterate over
+// the range could be either view based range or height based range.
+// when specifying the range, the start and end are inclusive, and the end must be greater than or
+// equal to the start
+type IterateJob struct {
+	Start uint64 // the start of the range
+	End   uint64 // the end of the range
+}
+
+// IterateJobCreator is an interface for creating iterate jobs
+type IteratorJobCreator interface {
+	// CreateJob takes a progress reader which is used to read the progress of the iterator
+	// and returns an iterate job that specifies the range of blocks to iterate over
+	CreateJob(IterateProgressReader) (IterateJob, error)
+}
+
+// IterateProgressReader reads the progress of the iterator, useful for resuming the iteration
+// after restart
+type IterateProgressReader interface {
+	// ReadNext reads the next block to iterate
+	// caller must ensure the reader is created by the IterateProgressInitializer,
+	// otherwise ReadNext would return exception.
+	ReadNext() (uint64, error)
+}
+
+// IterateProgressWriter saves the progress of the iterator
+type IterateProgressWriter interface {
+	// SaveNext persists the next block to be iterated
+	SaveNext(uint64) error
+}
+
+// IterateProgressInitializer is an interface for initializing the progress of the iterator
+// a initializer must be used to ensures the initial next block to be iterated is saved in
+// storage before creating the block iterator
+type IterateProgressInitializer interface {
+	Init() (IterateProgressReader, IterateProgressWriter, error)
+}
+
+// BlockIterator is an interface for iterating over blocks
+type BlockIterator interface {
+	// Next returns the next block in the iterator
+	// Note: a block will only be iterated once in a single iteration, however
+	// if the iteration is interrupted (e.g. by a restart), the iterator can be
+	// resumed from the last checkpoint, which might result in the same block being
+	// iterated again.
+	Next() (blockID flow.Identifier, hasNext bool, exception error)
```

Comment:
what do you think about making this a [go iterator](https://pkg.go.dev/iter#hdr-Iterators)? Instead of `Next()` it could be

```golang
func (b *HeightIterator) Range() iter.Seq2[flow.Identifier, error] {
	return func(yield func(flow.Identifier, error) bool) {
		for b.nextHeight <= b.endHeight {
			next, err := b.headers.BlockIDByHeight(b.nextHeight)
			if err != nil {
				yield(flow.ZeroID, fmt.Errorf("failed to fetch block at height %v: %w", b.nextHeight, err))
				return
			}
		
			b.nextHeight++
		
			if !yield(next, nil) {
				return
			}
		}
	}
}
```

then
```golang
for blockID, err := range heightIterator.Range() {
	...
}
```

Reply:
I like this idea!

Reply:
This is a go 1.23 feature right? 1.22 seems doesn't have it yet. 

Thanks, I will add a TODO once we upgrade to 1.23

Reply:
hopefully it will be soon: https://github.com/onflow/flow-go/pull/6922